testsuite/model/src/test/java/org/keycloak/testsuite/model/infinispan/CacheExpirationTest.java

```diff
@@ -51,69 +51,75 @@
                 } while (! cache.getAdvancedCache().getDistributionManager().isJoinComplete());
-                cache.keySet().forEach(s -> {});
-            });
-            String site = CONFIG.scope("connectionsInfinispan", "default").get("siteName");
-            log.debug("Cluster joined " + site);
-
-            // access the items in the local cache in the different site (site-2) in order to fetch them from the remote cache
-            if ("site-2".equals(site)) {
-                inComittedTransaction(session -> {
-                    InfinispanConnectionProvider provider = session.getProvider(InfinispanConnectionProvider.class);
-                    Cache<String, Object> cache = provider.getCache(InfinispanConnectionProvider.WORK_CACHE_NAME);
-                    cache.get("1-2");
-                    cache.get("1-2-3");
-                });
-            }
-            int c = getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class);
-            maxCountOfInstances.getAndAccumulate(c, Integer::max);
-            assumeThat("Seems we're running on a way too slow a computer", System.currentTimeMillis() - putTime.get(), Matchers.lessThan(20000L));
-
-            // Wait for at most 3 minutes which is much more than 15 seconds expiration set in DefaultInfinispanConnectionProviderFactory
-            for (int i = 0; i < 3 * 60; i++) {
-                try { Thread.sleep(1000); } catch (InterruptedException ex) { Thread.currentThread().interrupt(); throw new RuntimeException(ex); }
-                if (getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class) == 0) {
-                    break;
-                }
-            }
 
-            c = getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class);
-            minCountOfInstances.getAndAccumulate(c, Integer::min);
+                String site = CONFIG.scope("connectionsInfinispan", "default").get("siteName");
+                log.debug("Cluster joined " + site);
+
+                log.debug("Waiting for cache to receive the two elements within the cluster");
+                do {
+                    try { Thread.sleep(1000); } catch (InterruptedException ex) { Thread.currentThread().interrupt(); throw new RuntimeException(ex); }
+                } while (cache.entrySet().stream()
+                        .filter(me -> me.getValue() instanceof AuthenticationSessionAuthNoteUpdateEvent)
+                        .count() != 2);
+
+                assertThat(cache.get("1-2"), notNullValue());
+                assertThat(cache.get("1-2-3"), notNullValue());
+
+                log.debug("Waiting two elements to expire");
+                do {
+                    try { Thread.sleep(1000); } catch (InterruptedException ex) { Thread.currentThread().interrupt(); throw new RuntimeException(ex); }
+                } while (cache.entrySet().stream()
```

Comment:
Iteration of cache is exactly what must be avoided - the test is to check that infinispan in keycloak does not leak any objects due to its invalid internal settings. Cache iteration makes expired items expired immediately but explicitly. We need to ensure valid behaviour of infinispan. If the test is unstable then we might need to check with infinispan team if there is any wrong settings.

Reply:
@hmlnarik - what you see here is a refactoring to understand what it was doing when analyzing the problem when caches weren't starting up.

I updated the commit so I don't call the cache.

Reply:
This change is now no longer part of this PR, it will be part of another PR.